cert/aziot-certd/src/lib.rs

```diff
@@ -195,654 +211,379 @@ impl UpdateConfig for Api {
+            for (name, value) in fields {
+                builder
+                    .append_entry_by_text(name, value)
+                    .map_err(|err| Error::Internal(InternalError::CreateCert(Box::new(err))))?;
+            }
     }
+
+    Ok(builder.build())
 }
 
-fn create_cert<'a>(
+fn create_cert_inner<'a>(
     api: &'a mut Api,
     id: &'a str,
     csr: &'a [u8],
-    issuer: Option<(&'a str, &'a aziot_key_common::KeyHandle)>,
-) -> std::pin::Pin<Box<dyn std::future::Future<Output = Result<Vec<u8>, Error>> + Send + 'a>> {
+    issuer: Option<(&'a str, &'a KeyHandle)>
+) -> Pin<Box<dyn Future<Output = Result<Vec<u8>, BoxedError>> + Send + 'a>> {
+
     // Creating a cert is recursive in some cases. An async fn cannot recurse because its RPIT Future type would end up being infinitely sized,
     // so it needs to be boxed. So we have a non-async fn returning a boxed future, where the future being boxed is the result of an inner asyn fn,
     // and the recursive call is for the outer boxed-future-returning fn.
-
-    async fn create_cert_inner(
+    async fn recursion_trampoline(
         api: &mut Api,
         id: &str,
         csr: &[u8],
-        issuer: Option<(&str, &aziot_key_common::KeyHandle)>,
-    ) -> Result<Vec<u8>, Error> {
-        // Look up issuance options for this certificate ID.
+        issuer: Option<(&str, &KeyHandle)>
+    ) -> Result<Vec<u8>, BoxedError> {
         let cert_options = api.cert_issuance.certs.get(id);
 
-        if let Some((issuer_id, issuer_private_key)) = issuer {
-            // Issuer is explicitly specified, so load it and use it to sign the CSR.
-
-            let x509_req = openssl::x509::X509Req::from_pem(csr)
-                .map_err(|err| Error::invalid_parameter("csr", err))?;
-            let x509_req_public_key = x509_req
-                .public_key()
-                .map_err(|err| Error::invalid_parameter("csr", err))?;
-            if !x509_req
-                .verify(&x509_req_public_key)
-                .map_err(|err| Error::Internal(InternalError::CreateCert(Box::new(err))))?
-            {
-                return Err(Error::invalid_parameter(
-                    "csr",
-                    "CSR failed to be verified with its public key",
-                ));
+        if let Some(ref issuer) = issuer {
+            let expiry = &*Asn1Time::days_from_now(
+                    cert_options
+                        .and_then(|opts| opts.expiry_days)
+                        .unwrap_or(30)
+                )?;
+            let name_override = cert_options
+                .and_then(|opts| opts.subject.as_ref())
+                .map(build_name)
+                .transpose()?;
+
+            let req = X509Req::from_pem(csr)
+                .or_else(|_| X509Req::from_der(csr))?;
+            let pubkey = req.public_key()?;
+
+            if !req.verify(&pubkey)? {
+                Err("CSR failed to be verified with its public key".to_owned())?
```

Comment:
Parameter validation is now done before entering `create_cert_inner`: [reference](https://github.com/Azure/iot-identity-service/blob/3e44872269a9a6fe43e0e6e791332784f82eaa2a/cert/aziot-certd/src/lib.rs#L126-L146).

Reply:
>Parameter validation is now done before entering `create_cert_inner`: [reference](https://github.com/Azure/iot-identity-service/blob/6dc85fd27362b0013551bc39f633711bad41c84f/cert/aziot-certd/src/lib.rs#L127-L147).

Not true: https://github.com/Azure/iot-identity-service/pull/296/files#diff-ce1948fdb394a64e8a42f87d92e2057992c63a20520bad0c4c7359b64d0854c2R316 (cert/aziot-certd/src/lib.rs:316 in case it doesn't load.)

This kind of mistake is easy to make again in the future. Please change it back to returning `Error` so that every error return is consciously converted to the appropriate error kind.